extensions/binding/org.eclipse.smarthome.binding.mqtt.generic/src/main/java/org/eclipse/smarthome/binding/mqtt/generic/internal/discovery/AbstractMQTTDiscovery.java

```diff
@@ -0,0 +1,82 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.binding.mqtt.generic.internal.discovery;
+
+import java.util.Date;
+import java.util.Set;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.binding.mqtt.discovery.MQTTTopicDiscoveryParticipant;
+import org.eclipse.smarthome.binding.mqtt.discovery.MQTTTopicDiscoveryService;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+
+/**
+ * Base MQTT discovery class. Responsible for connecting to the {@link MQTTTopicDiscoveryService}.
+ *
+ * @author David Graeff - Initial contribution
+ */
+@NonNullByDefault
+public abstract class AbstractMQTTDiscovery extends AbstractDiscoveryService implements MQTTTopicDiscoveryParticipant {
+    final protected String subscribeTopic;
+
+    @NonNullByDefault({})
+    protected MQTTTopicDiscoveryService mqttTopicDiscovery;
+
+    public AbstractMQTTDiscovery(@Nullable Set<ThingTypeUID> supportedThingTypes, int timeout,
+            boolean backgroundDiscoveryEnabledByDefault, String baseTopic) {
+        super(supportedThingTypes, timeout, backgroundDiscoveryEnabledByDefault);
+        this.subscribeTopic = baseTopic;
+    }
+
+    public void setMultiConnectionTopicSubscribeService(MQTTTopicDiscoveryService service) {
```

Comment:
Since inheritance and OSGi services always suck a little bit what do you think about this:

Defining an abstract `getMQTTDiscoervyService` method here and implement it in the HomeassistantDiscovery and HomieDiscovery and hold the instance of the service there? This frees further implementors of calling `super.setMQTT....` and decouples this abstract class from the OSGi DS lifecycle.

Reply:
good idea. done